### PR TITLE
Add get_sink_index method to Python API.

### PR DESCRIPTION
**Pull request by kcin96**
### Details:
 - Adds `get_sink_index` method to the Python API. 
 - Adds tests.  

### Tickets:
 - 131043

This PR links issue #25116

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 1]
```diff
@@ -750,6 +750,71 @@ void regclass_graph_Model(py::module m) {
...
+                    :return: Index of sink node referenced by output handle.
+                    :rtype: int
+                  )");
+
+    model.def("get_sink_index",
+              [](ov::Model& self, const py::object& node) -> int64_t {
+                if (py::isinstance<ov::op::v6::Assign>(node)) {
+                    auto sink = std::dynamic_pointer_cast<ov::op::Sink>(node.cast<std::shared_ptr<ov::op::v6::Assign>>());
+                    int64_t pos = 0;
+                    for (const auto& s: self.get_sinks()) {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:786]: I think this cycle may be moved to separate function and reused in both overloadings.

Comment (kcin96) [reply src/bindings/python/src/pyopenvino/graph/model.cpp]: Ok, I've added a it to a separate function.

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 2]
```diff
@@ -750,6 +761,57 @@ void regclass_graph_Model(py::module m) {
                 :rtype: int
              )");
 
+    model.def(
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<ov::Node>& value) -> int64_t {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:766]: Could you please also add overload for `(ov::Model& self, const ov::Output<const ov::Node>& value)`

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 3]
```diff
@@ -128,6 +128,17 @@ static std::unordered_map<std::string, ov::PartialShape> get_variables_shapes(co
     return variables_shape_map;
 }
 
+static int64_t find_sink_position(const ov::SinkVector sinks, std::shared_ptr<ov::op::Sink> sink) {
```
Comment (praasz) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:131]:
```suggestion
static int64_t find_sink_position(const ov::SinkVector& sinks, const ov::op::Sink* const sink) {
```
or

```suggestion
static int64_t find_sink_position(const ov::SinkVector& sinks, const std::shared_ptr<ov::op::Sink>& sink) {
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 4]
```diff
@@ -750,6 +761,57 @@ void regclass_graph_Model(py::module m) {
                 :rtype: int
              )");
 
+    model.def(
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<ov::Node>& value) -> int64_t {
+            int64_t pos = -1;
+            if (ov::is_type<ov::op::v6::Assign>(value.get_node_shared_ptr())) {
+                auto sink = std::dynamic_pointer_cast<ov::op::Sink>(value.get_node_shared_ptr());
+                pos = find_sink_position(self.get_sinks(), sink);
```
Comment (praasz) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:770]:
```suggestion
                return find_sink_position(self.get_sinks(), sink);
```
`pos` can be remvoed.

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 2 above)
Comment (kcin96) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:777]: Sure, I've added the overload now.

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 5]
```diff
@@ -750,6 +772,81 @@ void regclass_graph_Model(py::module m) {
                 :rtype: int
              )");
 
+    model.def(
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<ov::Node>& value) -> int64_t {
+            int64_t pos = -1;
+            if (ov::is_type<ov::op::v6::Assign>(value.get_node_shared_ptr())) {
+                auto sink = std::dynamic_pointer_cast<ov::op::Sink>(value.get_node_shared_ptr());
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:780]:
```suggestion
            int64_t pos = -1;
            auto
            if (ov::is_type<ov::op::v6::Assign>(value.get_node_shared_ptr())) {
                auto sink = std::dynamic_pointer_cast<ov::op::Sink>(value.get_node_shared_ptr());
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 6]
```diff
@@ -750,6 +772,81 @@ void regclass_graph_Model(py::module m) {
...
+                return find_sink_position(self.get_sinks(), sink);
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:781]:
```suggestion
            int64_t pos = -1;
            auto node = value.get_node_shared_ptr();
            if (ov::is_type<ov::op::v6::Assign>(node)) {
                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<ov::op::Sink>(node));
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 7]
```diff
@@ -750,6 +772,81 @@ void regclass_graph_Model(py::module m) {
...
+                    :rtype: int
+                  )");
+
+    model.def(
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<const ov::Node>& value) -> int64_t {
+            int64_t pos = -1;
+            if (ov::is_type<ov::op::v6::Assign>(value.get_node_shared_ptr())) {
+                auto sink = std::dynamic_pointer_cast<const ov::op::Sink>(value.get_node_shared_ptr());
+                return find_sink_position(self.get_sinks(), sink);
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:805]:
```suggestion
            int64_t pos = -1;
            auto node = value.get_node_shared_ptr();
            if (ov::is_type<ov::op::v6::Assign>(node)) {;
                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<const ov::op::Sink>(node));
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 8]
```diff
@@ -128,6 +128,28 @@ static std::unordered_map<std::string, ov::PartialShape> get_variables_shapes(co
     return variables_shape_map;
 }
 
+static int64_t find_sink_position(const ov::SinkVector& sinks, const std::shared_ptr<ov::op::Sink>& sink) {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:131]:
```suggestion
template<typename T>
static int64_t find_sink_position(const ov::SinkVector& sinks, const std::shared_ptr<T>& sink)
```

Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:131]: I think it'd be better to have this function templated

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 9]
```diff
@@ -128,6 +128,29 @@ static std::unordered_map<std::string, ov::PartialShape> get_variables_shapes(co
...
+    for (const auto& s : sinks) {
+        if (s == sink) {
+            return pos;
+        }
+        pos++;
+    };
+    return -1;
+}
+
+static int64_t find_sink_position(const ov::SinkVector& sinks, const std::shared_ptr<const ov::op::Sink>& sink) {
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:143]: Please fix the `find_sink_position`.
Remove this one and check that everything works.

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 10]
```diff
@@ -750,6 +773,81 @@ void regclass_graph_Model(py::module m) {
...
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<const ov::Node>& value) -> int64_t {
+            int64_t pos = -1;
+            auto node = value.get_node_shared_ptr();
+            if (ov::is_type<ov::op::v6::Assign>(node)) {;
+                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<const ov::op::Sink>(node));
+            } else {
+                throw py::type_error("Incorrect argument type. Output sink node is expected as argument.");
+            }
+            return pos;
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:810]:
```suggestion
            auto node = value.get_node_shared_ptr();
            if (ov::is_type<ov::op::v6::Assign>(node)) {;
                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<const ov::op::Sink>(node));
            } else {
                throw py::type_error("Incorrect argument type. Output sink node is expected as argument.");
            }
```

File: src/bindings/python/src/pyopenvino/graph/model.cpp [hunk 11]
```diff
@@ -750,6 +773,81 @@ void regclass_graph_Model(py::module m) {
...
+        "get_sink_index",
+        [](ov::Model& self, const ov::Output<ov::Node>& value) -> int64_t {
+            int64_t pos = -1;
+            auto node = value.get_node_shared_ptr();
+            if (ov::is_type<ov::op::v6::Assign>(node)) {
+                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<ov::op::Sink>(node));
+            } else {
+                throw py::type_error("Incorrect argument type. Output sink node is expected as argument.");
+            }
+            return pos;
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/model.cpp:786]:
```suggestion
            auto node = value.get_node_shared_ptr();
            if (ov::is_type<ov::op::v6::Assign>(node)) {
                return find_sink_position(self.get_sinks(), std::dynamic_pointer_cast<ov::op::Sink>(node));
            } else {
                throw py::type_error("Incorrect argument type. Output sink node is expected as argument.");
            }
```

Comment (akuporos) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:786]: Looks like `pos` isn't need here anymore

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 9 above)
Comment (kcin96) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:143]: I've removed the second function and checked that it works.

File: src/bindings/python/src/pyopenvino/graph/model.cpp (hunk 11 above)
Comment (kcin96) [reply src/bindings/python/src/pyopenvino/graph/model.cpp:786]: I'll remove them.

**Comment by mlukasze** [comment]
build_jenkins

**Review by akuporos — APPROVED** [review]
LGTM,

Thank you for your contribution!

**Comment by akuporos** [comment]
build_jenkins

**Review by p-wysocki — APPROVED** [review]

**Comment by mlukasze** [comment]
thanks @kcin96!
Your change will be a part of our next release :)

**Pull request merged**